scwx-qt/source/scwx/qt/manager/marker_manager.cpp

```diff
@@ -207,57 +228,99 @@ size_t MarkerManager::marker_count()
    return p->markerRecords_.size();
 }
 
-std::optional<types::MarkerInfo> MarkerManager::get_marker(size_t index)
+std::optional<types::MarkerInfo> MarkerManager::get_marker(types::MarkerId id)
 {
    std::shared_lock lock(p->markerRecordLock_);
+   if (!p->idToIndex_.contains(id))
+   {
+      return {};
+   }
+   size_t index = p->idToIndex_[id];
    if (index >= p->markerRecords_.size())
    {
+      logger_->warn("id in idToIndex_ but out of range!");
       return {};
    }
    std::shared_ptr<MarkerManager::Impl::MarkerRecord>& markerRecord =
       p->markerRecords_[index];
    return markerRecord->toMarkerInfo();
 }
 
-void MarkerManager::set_marker(size_t index, const types::MarkerInfo& marker)
+std::optional<size_t> MarkerManager::get_index(types::MarkerId id)
+{
+   std::shared_lock lock(p->markerRecordLock_);
+   if (!p->idToIndex_.contains(id))
+   {
+      return {};
+   }
+   return p->idToIndex_[id];
+}
+
+void MarkerManager::set_marker(types::MarkerId id, const types::MarkerInfo& marker)
 {
    {
       std::unique_lock lock(p->markerRecordLock_);
+      if (!p->idToIndex_.contains(id))
+      {
+         return;
+      }
+      size_t index = p->idToIndex_[id];
       if (index >= p->markerRecords_.size())
       {
+         logger_->warn("id in idToIndex_ but out of range!");
          return;
       }
       std::shared_ptr<MarkerManager::Impl::MarkerRecord>& markerRecord =
          p->markerRecords_[index];
       markerRecord->markerInfo_ = marker;
    }
-   Q_EMIT MarkerChanged(index);
+   Q_EMIT MarkerChanged(id);
    Q_EMIT MarkersUpdated();
 }
 
 void MarkerManager::add_marker(const types::MarkerInfo& marker)
 {
+   types::MarkerId id;
    {
       std::unique_lock lock(p->markerRecordLock_);
+      id = p->NewId();
+      size_t index = p->markerRecords_.size();
+      p->idToIndex_.emplace(id, index);
       p->markerRecords_.emplace_back(std::make_shared<Impl::MarkerRecord>(marker));
+      p->markerRecords_[index]->markerInfo_.id = id;
    }
-   Q_EMIT MarkerAdded();
+   Q_EMIT MarkerAdded(id);
    Q_EMIT MarkersUpdated();
 }
 
-void MarkerManager::remove_marker(size_t index)
+void MarkerManager::remove_marker(types::MarkerId id)
 {
    {
       std::unique_lock lock(p->markerRecordLock_);
+      if (!p->idToIndex_.contains(id))
+      {
+         return;
+      }
+      size_t index = p->idToIndex_[id];
       if (index >= p->markerRecords_.size())
       {
+         logger_->warn("id in idToIndex_ but out of range!");
          return;
       }
 
       p->markerRecords_.erase(std::next(p->markerRecords_.begin(), index));
+      p->idToIndex_.erase(id);
+
+      for (auto& pair : p->idToIndex_)
+      {
+         if (pair.second > index)
+         {
+            p->idToIndex_[pair.first] = pair.second - 1;
```

Comment:
Minor: Since pair.second is a reference, can simplify here, `pair.second -= 1`?

Reply:
Seems to work.